Ambiguity warnings on definition -> error on usage

timholy opened the pull request:
'Nuf said: [EDIT: updated]

``` jl
julia> foo(x, y) = 1
foo (generic function with 1 method)

julia> foo(x::Int, y) = 2
foo (generic function with 2 methods)

julia> foo(x, y::Int) = 3                # look Ma, no annoying warning!
foo (generic function with 3 methods)

julia> foo(3.2, 4.1)                     # You can use foo without trouble...
1

julia> foo(3, 4)                         # ...except when the correct method is ambiguous
ERROR: MethodError: foo(::Int64, ::Int64) is ambiguous. Candidates:
  foo(x::Int64, y) at REPL[2]:1
  foo(x, y::Int64) at REPL[3]:1
 in eval(::Module, ::Any) at ./boot.jl:228

```

Fixes #6190.

~~I think the plan was to eventually turn these into errors, but that requires extraordinary confidence in the type-intersection machinery and won't happen until after the type system overhaul. In the meantime, I realized this might be an easy fix.~~

To help package developers pro-actively discover ambiguities, we might want to build some tools ([since added]) to scan method tables.

JeffBezanson commented:
Ha, that is a cleverly minimal fix. It would be great to have a change in this direction in the system ASAP.

It would really seal the deal if we could go back and remove the `ambig` references if a disambiguating definition is later added. That would fix #270, and who can resist closing a 3-digit issue?

JeffBezanson commented:
Also --- do we want to do this on every call, or only the first? Also might be good to print a stack trace, since uses are harder to locate than definitions.

StefanKarpinski commented:
You don't actually need to completely trust the type intersection machinery if you check for dispatch matches. Check if the actual argument types match more than one of the candidate methods when you think you *might* have an ambiguity. There are three cases:

1. No matches: raise a no method error.
2. One match: call the method.
3. Many matches: method ambiguity error.

I guess the hard part is knowing which methods you need to check for a match since currently, you just look for the first match when doing dispatch and this would require continuing beyond that.

JeffBezanson commented:
Good point --- this is all on the slow path anyway, so we could probably add that kind of logic.

timholy commented:
@JeffBezanson,

> It would really seal the deal if we could go back and remove the ambig references if a disambiguating definition is later added. That would fix #270, and who can resist closing a 3-digit issue?

Tempting indeed! There is one complication that arises because a method can be ambiguous with more than one other method. For example, if you define methods in this order:
```jl
julia> foo(x, y) = 1
julia> foo(x::Int, y) = 2
julia> foo(x, y::Int) = 3
julia> foo(x::Number, y::Int) = 4
```
then 2 and 3 point to each other as an ambiguous pair, while 4 points to 2. Then if you define this method,
```jl
julia> foo(x::Int, y::Int)
```
the edge between 2 and 3 gets severed, but 2 should then point to 4. To solve this, there are two possible options: (1) pass through the whole method list and look for new ambiguities, or (2) keep track of the whole "ambiguity graph" in the first place. Is this worth the performance cost?

> Also --- do we want to do this on every call, or only the first? 

I haven't played with this extensively, but caching seems to already handle a version of this for us:
```jl
julia> foo(x, y) = 1
foo (generic function with 1 method)

julia> foo(x::Integer, y) = 2
foo (generic function with 2 methods)

julia> foo(x, y::Integer) = 3
foo (generic function with 3 methods)

julia> foo(1, 2)
WARNING: Definition 
    foo(Integer, Any) at REPL[2]:1
is ambiguous with: 
    foo(Any, Integer) at REPL[3]:1
See the documentation for tips on eliminating ambiguities.
2

julia> foo(1,2)
2

julia> foo(0x01, 2)
WARNING: Definition 
    foo(Integer, Any) at REPL[2]:1
is ambiguous with: 
    foo(Any, Integer) at REPL[3]:1
See the documentation for tips on eliminating ambiguities.
2

julia> foo(0x01, 2)
2

julia> foo(1, 0x02)
WARNING: Definition 
    foo(Integer, Any) at REPL[2]:1
is ambiguous with: 
    foo(Any, Integer) at REPL[3]:1
See the documentation for tips on eliminating ambiguities.
2

julia> foo(1, 0x02)
2

julia> @noinline bar(x, y) = foo(x, y)
bar (generic function with 1 method)

julia> bar(1, 2)
2
```

> Also might be good to print a stack trace, since uses are harder to locate than definitions.

Is it a use-site problem, though? I'm not fully convinced, although I guess maybe you could say it is, in the same way that a `MethodError` could potentially be a use-site problem. But for cases where you're inserting this into a method you're compiling, won't that mean that we end up compiling the backtrace into the function? In cases where type intersection just came to the wrong conclusion, that seems likely to cause a pretty catastrophic performance hit.

Oh, wait, maybe there is a solution: what if we have codegen print the warning, saying "while compiling foo at src/myfile.jl: 17, an ambiguity in blah blah was detected." Where would such a warning get inserted? Presumably somewhere in `inference.jl`?

@StefanKarpinski,

> I guess the hard part is knowing which methods you need to check for a match since currently, you just look for the first match when doing dispatch and this would require continuing beyond that.

I was nervous about that. Isn't it essentially predictable that this will lead to a 2-fold slowdown in method lookup? If you basically assume that, on average, we traverse half the method table. I don't have a great sense for how much method lookup adds to compile time/runtime; if it's only a small fraction, then this might be worth it.

JeffBezanson commented:
Hmm, this is indeed getting more complicated as we poke at it.

I don't think the warning should depend on cache behavior (sometimes there is a new entry for new concrete types, sometimes not), or on when code is compiled. We also need to be sure to find at least as many ambiguities as we do now, while excluding those that are resolved by subsequent definitions.

It actually seems easier to give an error (when the call actually happens) instead of a warning, since then if we discover the problem at compile time the compiler can just insert a call to `throw`, which is quite easy to do.

The easiest way to go about this is to look for ambiguities on every slow-path method lookup. Although that's not a hot path, the search might be expensive enough to make a dent. Maybe we could do something like using the current check to set a "might be ambiguous" bit as a hint.

StefanKarpinski commented:
+1 to making it an error, not just a warning.

JeffBezanson commented:
Dusting off and putting on my language-lawyer hat, if you define "matching method" as "unique most specific matching method", then this is literally a no-method error.

timholy commented:
With regards to making it an error, I'm worried about:
#10174
#9588
#6048
#5460
(These are just the open issues, I didn't check closed ones.) In many/all of these cases, it is actually dispatching to the correct method. For this reason, I think we can't make it throw an error just yet.

Aside from that point, there seem to be two issues to resolve:
- Only warn once: fixed in bc6bc4dcc2b1db40415888e5f880a7b247fa6d39.
- Provide line info from "call site" (during compilation). See if you can stomach the solution in 23845b56aa573091fb5164b87b4a8d0e37ae5fc7. It's only a sketch, the source line info is very incomplete, but I'm not sure how to get to the line number without passing `lno` all the way down (or using a global).

toivoh commented:
Since it's such a small change, would it be possible to keep the definition time warnings with a switch for those of us who still like them?

timholy commented:
I definitely understand wanting that. However, just to ask: would an `ambiguities(MyModule, Base, ...)` function be equally attractive? Something that would scan for any ambiguities in a list of modules. (I'm not promising it's possible, but I'd guess it is.) A potential advantage is that we do end up deciding to solve #270, then it won't make much sense to do this any other way except, e.g., at module closure or upon user request.

tkelman commented:
I'd be tempted to add a default flag to `Pkg.test`, like we have for coverage, if this ambiguity detection API only ran when asked.

toivoh commented:
I would like it to be easy to set up my Julia installation to warn me about method ambiguities as soon as I make them, without me having to explicitly ask.

Then we'll see how long that will be workable, if no one else is getting the warnings then I might just be flooded with warnings. Then I guess that I would have to go with a more selective approach. But I would still prefer to get the warnings by default.

timholy commented:
"As soon as you make them" means "definition order matters," which it won't if we end up fixing #270. Are you saying you'd prefer to keep the order-sensitive behavior? Or just that you don't want to have to remember to ask for ambiguity detection. Combining your request for making it easy with @tkelman's comment, we could conceivably have a flag that would cause `check_ambiguities(mod, Base)` to run whenever a module is closed.

However, I'm noting that Stefan's proposal---eliminate ambiguity-detection entirely from the process of method definition, and only check on method lookup---seems to me like the only plausible way to solve #5460. In such cases, ambiguity might no longer be abstractly decidable, and you might not be able to have a `check_ambiguities` function at all. (Or it might give some false positives.)

timholy commented:
Actually, I no longer think those issues I linked above are a deal breaker for Stefan's proposal: of the 4, 1 was stochastic and may no longer happen, 1 should clearly fall out of a "morespecific" analysis, and the remaining 2 have to do with problems our "morespecific" analysis has with varargs. But if you're looking up a specific call signature, presumably that effectively resolves the varargs, and so I'm guessing "morespecific" will work properly.

toivoh commented:
Just don't want to have to remember to ask for ambiguity detection. I would be fine with a few spurious warnings a la #5460 in that case.

StefanKarpinski commented:
> Dusting off and putting on my language-lawyer hat, if you define "matching method" as "unique most specific matching method", then this is literally a no-method error.

These can certainly both be `MethodError`s but they should definitely give different error messages, otherwise it's going to be rather confusing if no matching method and too many matching methods are printed the same.

timholy commented:
OK, I've reworked this. It now throws an error when you try to call an ambiguous method. The overall architecture is basically along the lines of the discussion above:
- Register all potential ambiguities at function-definition time
- At compile/call time, check whether any of the potentially-ambiguous method signatures intersect with the argument tuple-type. If so, throw an error.

This also solves #270. However, there is no attempt to try to go back and remove potential ambiguities from the "registrations" if a more specific method is defined later; instead, at compile time the more specific one gets selected, when appropriate, because it's earlier in the method table---the question of ambiguity simply never comes up. Easy.

However, there's still some work to be done:
- The error-throwing is just a stub; are there some good examples I can look at to see how to get codegen to insert an error call into the compiled code?
- To be safe, I find that I have to use `@noinline` in the function call: for really trivial signatures like the ones in the new `test/ambiguous.jl`, they are inlined even before calling `inlineable` in `inference.jl`. I spent some time trying to figure out where this happens, but haven't yet been successful (pointers appreciated).
- Need tools to scan for ambiguities as an aid to developers

tkelman commented:
Can we hold off on merging this until after we branch for release-0.5? I feel like this will have wide ranging consequences that we don't understand yet.

timholy commented:
I'm usually pretty cautious about these things, but here I disagree; I suspect this is fairly safe, in many ways safer than what we have now (warn, but then pick a random function to execute). It seems like the kind of behavior change we want to come with the .0 release of 0.5.

An important part of testing, however, will be for people who use packages that produce a slew of ambiguity warnings when used in combination (e.g., Images and DataArrays) to give it a whirl.

vtjnash commented on src/typemap.c:
```diff
@@ -623,9 +623,10 @@ jl_typemap_entry_t *jl_typemap_assoc_by_type(union jl_typemap_t ml_or_cache, jl_
 }
 
-jl_lambda_info_t *jl_typemap_assoc_exact(union jl_typemap_t ml_or_cache, jl_value_t **args, size_t n, int8_t offs)
+jl_lambda_info_t *jl_typemap_assoc_exact(union jl_typemap_t ml_or_cache, jl_value_t **args, size_t n, int8_t offs, jl_typemap_entry_t **m)
```
instead of adding an extra parameter, this can just return the `jl_typemap_entry_t*`

timholy replied:
I wondered about that. Will change.

timholy replied:
For a `jl_typemap_entry_t *m`, can I just set `m->func.linfo` with the value of any specialized `jl_lambda_info_t` I create, and thereby return both? Or would that tank other necessary information (e.g., `func.method`)?

I don't really even understand what a "guard entry" is, so this is probably a pretty naive question.

timholy replied:
Wait, I think I figured it out; that's valid for methods in the `mt->cache`, but not in `mt->defs`.

vtjnash replied:
yeah, "guard entries" isn't really a function of the TypeMap, instead it's a placeholder used by the cache logic using a value of NULL to signal when a there would be a match, but that the LambdaInfo for that match hasn't been created yet.

timholy committed:
Move ambiguity warnings from definition to usage: register ambiguities

For now, this disables any kind of ambiguity warning/error

timholy committed:
Throw error for an ambiguous call

timholy committed:
Refactor: methods returns multiples in case of ambiguity

Also:
- eliminates a lot of the dedicated error-handling for ambiguous methods in favor of returning NULL
- causes `precompile` to return true/false upon success/failure (useful for testing)

timholy commented on test/ambiguous.jl:
```diff
@@ -0,0 +1,59 @@
+
+# Test that ambiguous fail appropriately
+@test_throws AmbiguousError precompile(ambig, (UInt8, Int))
+# @test_throws AmbiguousError cfunction(ambig, Int, (UInt8, Int))
```
This is the only weakness I know of: if you call `cfunction` twice on an ambiguous call, you get a segfault. Haven't yet had time to track this down.

timholy commented:
This seems pretty much done, except for tools to scan for potential ambiguities (which can probably be a separate PR).

tkelman commented on base/base.jl:
```diff
@@ -34,6 +34,11 @@ type MethodError <: Exception
     args
 end
 
+type AmbiguousError <: Exception
```
Ambiguous or Ambiguity? Most of the other exception types are nouns not adjectives

hayd replied:
Would it be better to be more long-winded: `AmbiguousMethodError <: MethodError` ? (and `NoMethodError <: MethodError`.)

toivoh replied:
Well, it's not the error that's ambiguous anyway, so both
AmbiguousMethodError and AmbiguityError seem better than AmbiguousError.

timholy commented on src/dump.c:
```diff
@@ -841,6 +841,7 @@ static void jl_serialize_value_(ios_t *s, jl_value_t *v)
         write_int32(s, m->line);
         jl_serialize_value(s, (jl_value_t*)m->sig);
         jl_serialize_value(s, (jl_value_t*)m->tvars);
+        jl_serialize_value(s, (jl_value_t*)m->ambig);
```
I'm also a little uncertain about whether one should add this, or whether there's a scan for ambiguities upon deserialization.

JeffBezanson replied:
We probably want to keep this, to avoid re-computing ambiguities.

timholy commented on src/gf.c:
```diff
@@ -681,26 +682,33 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
+    else {
+        /* if (ambiguities != jl_nothing) */
+        /*     nf = NULL; */
+        /* else */
```
I wasn't certain whether `cache=1` means "please try to insert into the cache," or whether it means "you darn well better insert into the cache" (in which case returning `nf=NULL` seems like the better option).

timholy commented:
Tests pass locally. Renaming as proposed by @hayd done. (`MethodError` has to be deprecated, though, it can't be re-purposed right away as an abstract type.) Docs and NEWS added.

tkelman commented:
It doesn't have to be within this same PR, but we will need ambiguity detection tools running automatically on Base before many changes get made with this in place. Ambiguous definitions that creep into Base are almost always bugs, but I doubt many of them would trigger the errors here since that requires a level of type-combination test coverage that we don't have and can't measure.

timholy committed:
Add tools for testing for ambiguities

timholy commented:
...grumble grumble...of course he's right, TGFTK...grumble grumble...

OK, see if you like this. Note it's #270-satisfying, meaning it is independent of the order of method definition. As a bonus, it catches some "stale exports" that don't have corresponding objects defined anymore. Finally, now the standard tests will look for ambiguities in Base; we're clean now, and if an ambiguity is introduced in the future the tests should fail. So this represents another step forward.

I'll leave it up to others to determine how this gets run via `Pkg.test` or whatever.

JeffBezanson commented on base/inference.jl:
```diff
@@ -2281,6 +2281,17 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
     methsp = meth[2]
     method = meth[3]::Method
+    ambig = method.ambig
+    if ambig != nothing
```
I think handling this should be the responsibility of `ml_matches`. It can choose either to return no matches if there is definitely an ambiguity, or multiple matches if it isn't sure. Either way inlining will automatically be rejected.

JeffBezanson commented on base/reflection.jl:
```diff
@@ -170,10 +170,11 @@ end
 tt_cons(t::ANY, tup::ANY) = (@_pure_meta; Tuple{t, (isa(tup, Type) ? tup.parameters : tup)...})
 
-code_lowered(f, t::ANY=Tuple) = map(m -> (m::Method).lambda_template, methods(f, t))
+code_lowered(f, t::ANY=Tuple) = map(m -> (m::Method).lambda_template, checked_methods(f, t))
```
I believe this currently just returns an empty array instead of throwing an error.

JeffBezanson commented on src/codegen.cpp:
```diff
@@ -2736,30 +2752,39 @@ static jl_cgval_t emit_call(jl_expr_t *ex, jl_codectx_t *ctx)
+                ambiguities = jl_list_call_ambiguities_((jl_tupletype_t*)aty, m);
+                if (ambiguities != jl_nothing) {
+                    emit_ambiguous_call_error((jl_tupletype_t*)aty,
+                                          (jl_array_t*)ambiguities, ctx);
```
Indent?

JeffBezanson commented on src/gf.c:
```diff
@@ -681,26 +682,35 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
-        nf = cache_method(mt, &mt->cache, (jl_value_t*)mt, sig, tt, m, env);
+        nf = jl_get_specialized(m, sig, env);
+    else {
+        /* if (hasamb) */
```
?

timholy replied:
I wasn't (and still am not) sure whether it's acceptable to fail to insert a method into the cache when `cache = 1`. So I left this here to provoke commentary.

timholy replied:
For example, this function could return `NULL` if there are ambiguities.

JeffBezanson commented on src/gf.c:
```diff
@@ -1039,9 +1065,67 @@ jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types)
+{
+    if (m->ambig == jl_nothing)
+        return jl_nothing;
+    // Return a list of ambiguities Any[file1, line1, file2, line2, ...],
```
Would it make sense just to return an array of Method objects?

JeffBezanson commented on src/gf.c:
```diff
@@ -1039,9 +1065,67 @@ jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types)
 JL_DLLEXPORT void jl_compile_hint(jl_tupletype_t *types)
 {
-    (void)jl_get_specialization1(types);
+    (void)jl_checked_specialization1(types);
```
I think this function should continue not to throw errors. Also, the callers of `jl_get_specialization1` expect it to return NULL if anything goes wrong instead of throwing (if any callers don't check for NULL, they should). Particularly in codegen.cpp we should avoid throwing exceptions.

JeffBezanson commented on src/gf.c:
```diff
@@ -1504,6 +1588,20 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
+        mfunc = entry->func.linfo;
+    }
+    jl_method_t *m = mfunc->def;
+    if (m->ambig != jl_nothing) {
```
This code should be moved to the "cache miss case" above.

JeffBezanson commented on src/gf.c:
```diff
@@ -1504,6 +1588,20 @@ JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
+        ambiguities = jl_list_call_ambiguities_(tt, m);
+        if (ambiguities != jl_nothing) {
+            JL_GC_POP();
+            jl_ambiguous_method_error((jl_value_t*)tt, ambiguities);
```
I think `AmbiguousMethodError` should have the same fields as `NoMethodError`. Constructing the full list of ambiguities seems like a job for error printing. At an exception throwing site I prefer to quickly package up the minimum information you need. If the error is going to be printed, there's plenty of time to dig through method tables.

JeffBezanson commented:
Impressive, @timholy ! Not many have the guts to touch `jl_apply_generic` :)

I'm ok either way, but I think we have the option of keeping `MethodError` as it is. The extra clarification in the name might not be worth the amount of churn.

timholy commented:
Thanks for the review!

Combining several of your comments, I might try going back to having just plain `MethodError`, and see if I can disambiguate whether it's "no method" vs. "too many methods" at printing time. Is that crazy?

JeffBezanson commented:
That's an interesting idea --- given just a function and arguments, you can indeed figure out which problem it was. I kind of like the simplicity of having fewer exception types.

timholy commented:
There's also only on `NULL` a C function can return. So I'll try that and see if I encounter road blocks.

timholy committed:
Add docs and NEWS

timholy commented:
OK, ready again. Approximately 80 fewer added lines this way, I like it.

JeffBezanson commented on base/test.jl:
```diff
@@ -823,4 +824,44 @@ function test_approx_eq_modphase{S<:Real,T<:Real}(
+                        if m.ambig != nothing
+                            for m2 in m.ambig
+                                if Base.isambiguous(m, m2)
+                                    push!(ambs, (f, sortdefs(m.file, m.line, m2.file, m2.line)...))
```
Seems like it would be more elegant to return an array of 2-tuples of Methods.

vtjnash commented:
I think the main concern I have right now is that `(::Type{#T<:Any})(Any) at sysimg.jl:50` is currently ambiguous with at least 3 constructor methods in base (and one in the tests) but the detection for them is currently lacking. This can be seen here: https://travis-ci.org/JuliaLang/julia/jobs/126188064. (I think I may have understand how to fix the missing ambiguity issue, which will probably cause merge conflicts with this, but shouldn't conflict with the design.) This PR would make the detection of those error less noticeable (although hopefully with your PR it would now fail the tests).

JeffBezanson commented:
Fortunately, we're hoping to be able to remove that definition.

timholy commented:
This certainly doesn't do anything to address cases where an ambiguity is missed. But given we have a problem with that now anyway, isn't that orthogonal to this PR?

timholy commented:
OK, I'll merge this once CI passes.

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -7305,7 +7305,7 @@ digits!
     MethodError(f, args)
 
-A method with the required type signature does not exist in the given generic function.
+A method with the required type signature does not exist in the given generic function. Alternatively, there is no unique most-specific method.
```
run `make docs` and commit the rst

tkelman commented on base/test.jl:
```diff
@@ -823,4 +824,43 @@ function test_approx_eq_modphase{S<:Real,T<:Real}(
+    for mod in mods
+        for n in names(mod, true, imported)
+            try
+                f = eval(mod, n)
```
is this possible to do without using `eval`? especially if we're considering putting this in `Pkg.test` in some form, this seems like a vector for abuse.

yuyichao replied:
`getfield`

tkelman commented on base/test.jl:
```diff
@@ -823,4 +824,43 @@ function test_approx_eq_modphase{S<:Real,T<:Real}(
+    detect_ambiguities(mod1, mod2...)
+
+Test for ambiguous pairs of methods among the listed modules. Returns
+an array-of-`(Method,Method)` tuples.
```
just "array of ..." without the dashes would be cleaner imo

JeffBezanson commented on src/gf.c:
```diff
@@ -1039,9 +1056,35 @@ jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types)
 
-JL_DLLEXPORT void jl_compile_hint(jl_tupletype_t *types)
+// Specialize, but return NULL if there is a call ambiguity
+jl_lambda_info_t *jl_checked_specialization1(jl_tupletype_t *types)
```
I think this should be merged with `jl_get_specialization1`. I don't think there is a use case for getting a specialization that would be a method error (due to ambiguities) if actually called.

timholy replied:
I'm struggling with this a bit. `codegen.cpp` sometimes gets `NULL` back from `jl_get_specialization1` for other reasons. One example is this expression:
```jl
Expr(:call, top(:convert), Expr(:call, top(:typeof), SSAValue(28))::DataType, 0)::Any
```
which happens right after building `inference.jl`.

Presumably, I don't want to throw a `MethodError` in such cases, but if we merge `checked_specialization1` into `get_specialization1` then there won't be a way to know why you got `NULL` (and at least the way I've been doing this, that seems key to knowing when the emit the `MethodError`).

JeffBezanson replied:
Actually, codegen.cpp currently never emits MethodErrors directly. It just falls through to calling jl_apply_generic, and lets that handle throwing the error. So this change to codegen is totally separate and I believe not needed for this.

timholy replied:
Thanks. Works. Simpler and simpler...

JeffBezanson commented:
Thanks, @timholy. This is starting to look very good.

JeffBezanson commented:
:tada: 

Don't forget to close #270 too!

tkelman commented:
What are the "Skipping" messages about in this test?

timholy commented:
`<|` seems to be exported but not defined. There's also something interesting about how it gets parsed:
```jl
julia> Base.<|
ERROR: MethodError: no method matching .<(::Module, ::Base.#|)
Closest candidates are:
  .<(::Any, ::AbstractArray{T,N})
  .<(::AbstractArray{T,N}, ::Any)
 in eval(::Module, ::Any) at ./boot.jl:228
```
Once I discovered that, I just assumed it was true for the rest. But now I see that the two non-operators (`cluster_manager` and `active_repl`) are use-dependent---they are defined as globals but not set until something triggers them to be defined.

```jl
julia> Base.cluster_manager
ERROR: UndefVarError: cluster_manager not defined
 in eval(::Module, ::Any) at ./boot.jl:228
```

tlnagy commented:
Is this going to be backported to 0.4? Since a bunch of things were excised from Base for 0.5, but are still present in 0.4, this leads all kinds of ambiguity warnings on 0.4.

For example, several packages (like HypothesisTests) depend on `Combinatorics.jl` which throws a huge number of ambiguity warnings when imported on 0.4:

```julia
julia> using Combinatorics
WARNING: Method definition levicivita(AbstractArray{#T<:Integer, 1}) in module Base at combinatorics.jl:611 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:188.
WARNING: Method definition partitions(Integer) in module Base at combinatorics.jl:252 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/partitions.jl:26.
WARNING: Method definition partitions(Integer, Integer) in module Base at combinatorics.jl:318 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/partitions.jl:93.
WARNING: Method definition partitions(AbstractArray{T<:Any, 1}) in module Base at combinatorics.jl:380 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/partitions.jl:158.
WARNING: Method definition partitions(AbstractArray{T<:Any, 1}, Int64) in module Base at combinatorics.jl:447 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/partitions.jl:228.
WARNING: Method definition permutations(Any) in module Base at combinatorics.jl:219 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:24.
WARNING: Method definition nthperm!(AbstractArray{T<:Any, 1}, Integer) in module Base at combinatorics.jl:70 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:136.
WARNING: Method definition prevprod(Array{Int64, 1}, Any) in module Base at combinatorics.jl:565 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/partitions.jl:354.
WARNING: Method definition combinations(Any, Integer) in module Base at combinatorics.jl:182 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/combinations.jl:42.
WARNING: Method definition nthperm(AbstractArray{#T<:Integer, 1}) in module Base at combinatorics.jl:92 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:161.
WARNING: Method definition nthperm(AbstractArray{T<:Any, 1}, Integer) in module Base at combinatorics.jl:89 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:157.
WARNING: Method definition factorial(#T<:Integer, #T<:Integer) in module Base at combinatorics.jl:56 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/factorials.jl:18.
WARNING: Method definition factorial(Integer, Integer) in module Base at combinatorics.jl:66 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/factorials.jl:28.
WARNING: Method definition parity(AbstractArray{#T<:Integer, 1}) in module Base at combinatorics.jl:642 overwritten in module Combinatorics at /Users/tamasnagy/.julia/v0.4/Combinatorics/src/permutations.jl:221.
```

tkelman commented:
No, this won't be backported. Report issues in those packages. Check first whether they've been fixed on master or if there are already open issues/PR's regarding ambiguity warnings.

mbauman commented:
Note that these *aren't* ambiguity warnings — Combinatorics is actually redefining those methods in base. Combinatorics could probably add a version check to ensure it only overrides the Base stubs in versions where they're errors that are designed to be overridden.
